app/gui2/src/components/CodeEditor.vue

```diff
@@ -157,13 +161,136 @@ watchEffect(() => {
+          {
+            needsRefresh(update) {
+              return (
+                update.state.field(diagnosticsVersion) !==
+                update.startState.field(diagnosticsVersion)
+              )
+            },
+          },
+        ),
       ],
     }),
   )
+  viewInitialized.value = true
+})
+
+function changeSetToTextEdits(changes: ChangeSet) {
+  const textEdits = new Array<TextEdit>()
+  changes.iterChanges((from, to, _fromB, _toB, insert) =>
+    textEdits.push({ range: [from, to], insert: insert.toString() }),
+  )
+  return textEdits
+}
+function textEditToChangeSpec({ range: [from, to], insert }: TextEdit) {
+  return { from, to, insert }
+}
+
+let pendingChanges: ChangeSet | undefined
+let currentModule: MutableModule | undefined
+/** Set the editor contents the current module state, discarding any pending editor-initiated changes. */
+function resetView() {
+  console.info(`Resetting the editor to the module code.`)
+  pendingChanges = undefined
+  currentModule = undefined
+  const viewText = editorView.state.doc.toString()
+  const code = graphStore.moduleSource.text
+  editorView.dispatch({
+    changes: textChangeToEdits(viewText, code).map(textEditToChangeSpec),
+    annotations: synchronizedModule.of(graphStore.startEdit()),
+  })
+}
+
+/** Apply any pending changes to the currently-synchronized module, clearing the set of pending changes. */
+function commitPendingChanges() {
+  if (!pendingChanges || !currentModule) return
+  try {
+    currentModule.applyTextEdits(changeSetToTextEdits(pendingChanges), graphStore.viewModule())
+    graphStore.commitEdit(currentModule, undefined, 'local:CodeEditor')
+  } catch (error) {
+    console.error(`Code Editor failed to modify module`, error)
+    resetView()
+  }
+  pendingChanges = undefined
+}
+
+function updateListener() {
+  const debouncer = createDebouncer(0)
+  return EditorView.updateListener.of((update) => {
+    for (const transaction of update.transactions) {
+      const newModule = transaction.annotation(synchronizedModule)
+      if (newModule) {
+        // Flush the pipeline of edits that were based on the old module.
+        commitPendingChanges()
+        currentModule = newModule
+      } else if (transaction.docChanged && currentModule) {
+        pendingChanges = pendingChanges
+          ? pendingChanges.compose(transaction.changes)
+          : transaction.changes
+        // Defer the update until after pending events have been processed, so that if changes are arriving faster than
+        // we would be able to apply them individually we coalesce them to keep up.
+        debouncer(commitPendingChanges)
+      }
+    }
+  })
+}
+
+let needResync = false
+// Indicates a change updating the text to correspond to the given module state.
+const synchronizedModule = Annotation.define<MutableModule>()
+watch(
+  viewInitialized,
+  (ready) => {
+    if (ready) graphStore.moduleSource.observe(observeSourceChange)
+  },
+  { immediate: true },
```

Comment:
If `viewInitialized` is always set asynchronously, then this `immediate` has no real effect, since `ready` will always be false on the first execution.

Reply:
Actually in practice `projectStore.module` is already non-null when the top-level `watchEffect` above runs, so this needs `immediate` to work.